packages/common/src/interfaces.ts

```diff
@@ -4,6 +4,8 @@ export type Payload = temporal.api.common.v1.IPayload;
 
 /** Type that can be returned from a Workflow `execute` function */
 export type WorkflowReturnType = Promise<any>;
+export type WorkflowUpdateType = (...args: any[]) => any;
```

Comment:
Let's discuss this a bit more before making it impossible to use a normal function as a handler. I don't have a firm position yet but I'm hesitant. Suppose a user simply wants to mutate and return some internal workflow state:

```ts
export async function workflow() {
  let internalState = '';
  wf.setHandler(myUpdate, (arg: string): string => {
    internalState += arg;
    return internalState;
  });
}
```

- Do we really want the compiler to deny them the ability to do so with a normal JS function? Might some users be annoyed/frustrated by the error in their IDE ("I just want to mutate a local variable and return it, why does that have to be done asynchronously?")
- It's intellectually confusing: there is in fact no technical reason to ban sync handlers (we'd be doing it for ergonomics/type simplicity), but if we ban it it will lead some people to think that sync handlers inherently don't make sense, or are bad design.
- A Signal handler can be sync, why do we want the inconsistency?

Reply:
To be exact, the `WorkflowUpdateType` is not constraining user's handler signature anywhere. It is only used internally for the `updateHandlers` map, in which case callers (ie. SDK's code) should anyway assume that the function is async and need to be awaited (awaiting a non-thenable is fine anyway).

IMHO We should have never exposed publicly `WorkflowSignalType` and `WorkflowQueryType`. I'd be inclined to unexpose them now; that would technically be a breaking change, but I expect very few users would be affected.

Anyway, back to the present point, `WorkflowUpdateType` and `WorkflowUpdateValidatorType` could be moved to `workflow/src/internals.ts`, not exported, and `WorkflowUpdateType` can be made to return a `Promise<...>`.

Reply:
To clarify this conversation, my comment above was really addressing this comment from @bergundy:

> Updates should always be async functions IMHO, it simplifies the signatures for handles.

Reply:
Now, @bergundy's claim and your arguments could still be stated with regard to the `setHandler()`'s argument type, which currently allow update handlers to be either async or sync functions, as shown here:

```
export type Handler<
  Ret,
  Args extends any[],
  T extends UpdateDefinition<Ret, Args> | SignalDefinition<Args> | QueryDefinition<Ret, Args>
> = T extends UpdateDefinition<infer R, infer A>
  ? (...args: A) => R | Promise<R>
  : T extends SignalDefinition<infer A>
  ? (...args: A) => void | Promise<void>
  : T extends QueryDefinition<infer R, infer A>
  ? (...args: A) => R
  : never;
```

I strongly think we should leave that definition as it is. I don't want to impose `Promise` return values here; quite the opposite, I would actually prefer to discourage async handlers, though probably not to the point of denying that entirely. Writing a correct async handler in a workflow context is not a trivial thing, and most devs don't figure out these difficulties until they get bit in production by some "mysterious issues".

Reply:
>  I would actually prefer to discourage async handlers, though probably not to the point of denying that entirely


Hang on, Update handlers can execute Activities and child Workflows, sleep, and wait for conditions, which clearly require the handler to be async. Are you saying you don't view the ability to do those things as essential to the Update feature?

Reply:
Yeah, I'm fine with having a sync handler, but the update definition should not include `Promise`:

```ts
const def = defineUpdate<string, [string]>('myUpdate');
```

Should support:

```ts
setHandler(def, async (input) => {
  return input;
})
```

Reply:
And the signature for the client methods should use `Promise<Awaited<T>>` for the result type just in case `T` (the result type) is a `Promise<something>`.

Reply:
>  the update definition should not include `Promise`

Definitely agree. (And that is already so -- there was a stray `Promise<Foo>` in my features repo PR which I've removed.)

Reply:
In case it's not clear: the current version of my branch aims to (a) support both normal functions and async functions as handlers, and (b) avoid pollution from framework-derived promises.